nemo/lightning/io/connector.py

```diff
@@ -145,18 +146,20 @@ def nemo_setup(self, model: pl.LightningModule, trainer: Optional[pl.Trainer] =
             pl.Trainer: The trainer configured with the model and strategy.
         """
         from nemo.lightning import MegatronStrategy, Trainer
-        from nemo.lightning._strategy_lib import megatron_lazy_init_context
+        from nemo.lightning._strategy_lib import megatron_cpu_init_context
 
         _trainer = trainer or Trainer(
-            devices=1, accelerator="cpu", strategy=MegatronStrategy(store_optimizer_states=False)
+            devices=1,
+            accelerator="cpu",
+            strategy=MegatronStrategy(ckpt_save_optimizer=False, always_save_context=True),
         )
-
+        _trainer.state.fn = TrainerFn.FITTING
         _trainer.strategy.connect(model)
         _trainer.strategy.setup_environment()
 
         if not model.state_dict():
-            _trainer.strategy.lazy_init = True
-            with _trainer.init_module(), megatron_lazy_init_context(model.config):
+            _trainer.strategy.lazy_init = False
+            with _trainer.init_module(), megatron_cpu_init_context(model.config):
```

Comment:
@ashors1 who would know what this should be? All I know for sure is that when I get to my convert checkpoint code, I have a model with some parameters on device `meta` which megatron dist checkpointing silently skips 😬

Reply:
@akoumpa do you have any insight?

Reply:
To be clear this "fixes" that problem with the meta device, but I think it might put the model on the GPU for whatever reason (some weights were still marked bf16, which I don't think is possible on cpu).

Reply:
@jstjohn Hmm, I would double check the checkpoint conversion code. My assumption is even if the source model uses `meta` in its state_dict that shouldn't matter because those would be overwritten by the imported checkpoint.

One thing that's missing is [here](https://github.com/NVIDIA/NeMo/blob/main/nemo/lightning/io/state.py#L164) to add an assert statement if any tensors is on meta device.

I would not change `megatron_lazy_init_context` here.

Reply:
I'll try putting it back. What was happening before was that 
1. The megatron dist checkpointing load step was not complaining
2. when I got to `self.nemo_save` I would run into the following:

```
    @pytest.mark.needs_gpu
    def test_nemo1_checkpoint_conversion(
        tmpdir: Path, geneformer_config: GeneformerConfig, cells: List[List[str]], seed: int = 42
    ):
        with megatron_parallel_state_utils.distributed_model_parallel_state(32):
            converter = GeneformerNeMo1LightningModuleConnector(nemo1_checkpoint_path)
            assert isinstance(converter.tokenizer, GeneTokenizer)
            diffs = compare_dataclasses(converter.config, geneformer_config)
            skip_fields = {"nemo1_ckpt_path", "return_only_hidden_states", "init_method", "output_layer_init_method"}
            filt = [d for d in diffs if d["field"] not in skip_fields]
            assert filt == []
            out_config = tmpdir / "out_config"
>           converter.apply(out_config)  # currently crashes in here during self.nemo_save(out_path, trainer)

sub-packages/bionemo-geneformer/tests/bionemo/geneformer/test_model.py:463: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
sub-packages/bionemo-llm/src/bionemo/llm/model/biobert/connector.py:71: in apply
    self.nemo_save(output_path, trainer)
3rdparty/NeMo/nemo/lightning/io/connector.py:176: in nemo_save
    trainer.save_checkpoint(output_path)
/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/trainer.py:1365: in save_checkpoint
    self.strategy.save_checkpoint(checkpoint, filepath, storage_options=storage_options)
3rdparty/NeMo/nemo/lightning/pytorch/strategies/megatron_strategy.py:648: in save_checkpoint
    self.checkpoint_io.save_checkpoint(checkpoint, filepath, storage_options=storage_options)
3rdparty/NeMo/nemo/lightning/io/pl.py:131: in save_checkpoint
    return dist_checkpointing.save(
3rdparty/Megatron-LM/megatron/core/dist_checkpointing/serialization.py:382: in save
    sharded_strategy.save(sharded_state_dict, checkpoint_dir)
3rdparty/Megatron-LM/megatron/core/dist_checkpointing/strategies/fully_parallel.py:86: in save
    return self.base_strategy.save(sharded_state_dict, checkpoint_dir)
3rdparty/Megatron-LM/megatron/core/dist_checkpointing/strategies/base.py:218: in save
    async_request = self.async_save(sharded_state_dict, checkpoint_dir)
3rdparty/Megatron-LM/megatron/core/dist_checkpointing/strategies/torch.py:651: in async_save
    ) = save_state_dict_async_plan(
3rdparty/Megatron-LM/megatron/core/dist_checkpointing/strategies/state_dict_saver.py:117: in save_state_dict_async_plan
    storage_writer.prepare_write_data(central_plan, planner)
3rdparty/Megatron-LM/megatron/core/dist_checkpointing/strategies/filesystem_async.py:118: in prepare_write_data
    tensor_data = [
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

.0 = <list_iterator object at 0x7115ec5af0a0>

    tensor_data = [
>       (item, planner.resolve_data(item).detach().to("cpu", non_blocking=True))
        for item in bucket
        if item.type != WriteItemType.BYTE_IO
    ]
E   NotImplementedError: Cannot copy out of meta tensor; no data!

3rdparty/Megatron-LM/megatron/core/dist_checkpointing/strategies/filesystem_async.py:119: NotImplementedError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /workspaces/bionemo-fw-ea/3rdparty/Megatron-LM/megatron/core/dist_checkpointing/strategies/filesystem_async.py(119)<listcomp>()
-> (item, planner.resolve_data(item).detach().to("cpu", non_blocking=True))
(Pdb) 
```

Reply:
My apply method looks like this: 
```python
def apply(self, output_path: Path) -> Path:
        """Save this nemo1 checkpoint to the desired output path in nemo2 format."""
        nemo1_path = str(self)  # self is a Path object
        with tarfile.open(nemo1_path, "r") as old_ckpt:
            ckpt_file = old_ckpt.extractfile("./model_weights.ckpt")
            old_weights = torch.load(ckpt_file)
        target = self.init()
        trainer = self.nemo_setup(target)
        target.trainer = trainer
        self.convert_state(old_weights, target)
        self.nemo_save(output_path, trainer)

        print(f"Converted NeMo1, model at {self} saved to {output_path}")

        teardown(trainer, target)
        del trainer, target

        return output_path
 ```

Reply:
And my convert state, which silently maybe left things as meta?
```python
    def convert_state(self, source: Dict[str, torch.Tensor], target: BioBertLightningModule) -> BioBertLightningModule:
        """Convert the input state_dict keys from nemo1 biobert to nemo2 biobert."""
        te_mapping = self.is_te_mapping(target)  # check for TE layers.
        target.module.cpu()
        new_state_dict_from_old = {}
        for k, v in source.items():
            new_key = nemo1_to_nemo2_biobert_key_mapping(k, new_model_prefix="", te_mapping=te_mapping)
            new_state_dict_from_old[new_key] = v
        target.module.load_state_dict(new_state_dict_from_old, strict=not te_mapping)
        return target
```